### PR TITLE
chore(ci): add deploy docs workflow

### PR DESCRIPTION
**Pull request by katebatura**
### Please read and mark the following check list before creating a pull request:

 - [X] I read and followed the [CONTRIBUTING.md](https://github.com/akveo/nebular/blob/master/CONTRIBUTING.md) guide.
 - [X] I read and followed the [New Feature Checklist](https://github.com/akveo/nebular/blob/master/DEV_DOCS.md#new-feature-checklist) guide.
 
 #### Short description of what this resolves:

Adds `deploy-docs` workflow which deploys docs on every master commit.
Updates `build-docs` script to reuse a built docs app from `gh-pages` branch instead of building the app again.

### CODE REVIEW

**Review by yggg — CHANGES_REQUESTED** [review]

File: tools/deploy-docs/build-docs.ts [hunk 1]
```diff
@@ -72,17 +78,68 @@ async function buildDocs(versions: Version[]) {
...
+    await remove(projectDir);
+  }
+}
+
+async function checkFileExists(file) {
+  return fs.promises
+    .access(file, fs.constants.F_OK)
+    .then(() => true)
+    .catch(() => false);
+}
```
[the remark this answers is not in the document]
Comment (katebatura) [reply tools/deploy-docs/build-docs.ts]: This function isn't used anymore.

File: .github/workflows/deploy-docs.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+name: Deploy docs
+on:
+  push:
+    branches:
+      - master
+
```
Comment (yggg) [inline .github/workflows/deploy-docs.yml:6]: Remove unnecessary new line

File: .github/workflows/deploy-docs.yml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  release:
+    types:
+      - published
+
```
Comment (yggg) [inline .github/workflows/deploy-docs.yml:10]: Remove unnecessary new line

File: tools/deploy-docs/versions.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 [
   {
-    "checkoutTarget": "de1ada462",
+    "checkoutTarget": "774803db",
```
Comment (yggg) [inline tools/deploy-docs/versions.json:3]: Could we change it to `8.0.x`?

File: tools/deploy-docs/build-docs.ts [hunk 5]
```diff
@@ -30,6 +34,13 @@ export interface Version {
   log(`Cloning ${REPO_URL} into ${MASTER_BRANCH_DIR}`);
   await runCommand(`git clone ${REPO_URL} ${MASTER_BRANCH_DIR}`, { cwd: WORK_DIR });
 
+  log('Checkout gh-pages to check for existing');
+  await copyToBuildDir(MASTER_BRANCH_DIR, GH_PAGES_DIR);
+  await checkoutVersion('gh-pages', GH_PAGES_DIR);
+
+  log('Search for commit hash files in gh-pages');
+  const builtVersions: { hash; path }[] = await checkBuiltVersions();
```
Comment (yggg) [inline tools/deploy-docs/build-docs.ts:42]:
```suggestion
  log('Get build versions from gh-pages branch');
  await copyToBuildDir(MASTER_BRANCH_DIR, GH_PAGES_DIR);
  await checkoutVersion('gh-pages', GH_PAGES_DIR);
  const builtVersions: { hash; path }[] = await checkBuiltVersions();
```

File: tools/deploy-docs/build-docs.ts [hunk 6]
```diff
@@ -58,31 +69,85 @@ function ensureSingleCurrentVersion(versions: Version[]) {
...
+      if (existsInDirectory) {
+        builtVersions.push({
+          hash: await readFile(hashFilePathInDirectory, 'utf8'),
+          path: join(GH_PAGES_DIR, file),
+        });
+      }
+    }
+  }
+
+  console.log('Built versions in gh-pages:', builtVersions);
```
Comment (yggg) [inline tools/deploy-docs/build-docs.ts:99]: Use `log` and move to the main function after `checkBuiltVersions` call.

File: tools/deploy-docs/build-docs.ts [hunk 7]
```diff
@@ -113,6 +178,10 @@ async function addVersionTs(version: Version, versionTsPath: string) {
   await writeFile(versionTsPath, source, 'utf8');
 }
 
+async function createFileWithCommitHash(path: string, projectDir: string) {
```
Comment (yggg) [inline tools/deploy-docs/build-docs.ts:181]:
```suggestion
async function addCommitHash(path: string, projectDir: string) {
```
To conform to already existing `addVersionNameToPackageJson` and `addVersionTs`.

File: tools/deploy-docs/build-docs.ts [hunk 8]
```diff
@@ -58,31 +69,85 @@ function ensureSingleCurrentVersion(versions: Version[]) {
...
-  await runCommand('npm ci', { cwd: projectDir });
-  await addVersionNameToPackageJson(version.name, join(projectDir, 'package.json'));
-  await addVersionTs(version, join(projectDir, 'package.json'));
-  await buildDocsApp(projectDir, version.path);
-  await copy(join(projectDir, 'docs/dist'), distDir);
-  await outputFile(join(distDir, 'assets/ghspa.js'), ghspaScript);
-
-  await remove(projectDir);
+
+  const { stdout: currentHash } = await promisify(exec)('git rev-parse HEAD', { cwd: projectDir });
```
Comment (yggg) [inline tools/deploy-docs/build-docs.ts:122]: Extract into getStdout function.

File: tools/deploy-docs/build-docs.ts [hunk 9]
```diff
@@ -58,31 +69,85 @@ function ensureSingleCurrentVersion(versions: Version[]) {
...
+
+  const correspondingVersion = builtVersions.find((item) => currentHash === item.hash);
```
Comment (yggg) [inline tools/deploy-docs/build-docs.ts:124]:
```suggestion
  const existInGhPages = builtVersions.find((item) => currentHash === item.hash);
```

File: tools/deploy-docs/build-docs.ts [hunk 10]
```diff
@@ -58,31 +69,85 @@ function ensureSingleCurrentVersion(versions: Version[]) {
...
+    await buildDocsApp(projectDir, version.path);
+    await createFileWithCommitHash(join(OUT_DIR, FILE_WITH_HASH), projectDir);
+    await copy(join(projectDir, OUT_DIR), distDir);
+    await outputFile(join(distDir, 'assets/ghspa.js'), ghspaScript);
+
+    await remove(projectDir);
+  }
+}
+
+async function copyExistingDocs(version, correspondingVersionPath, distDir) {
```
Comment (yggg) [inline tools/deploy-docs/build-docs.ts:141]:
```suggestion
async function copyFromGhPages(version, correspondingVersionPath, distDir) {
```

**Review by yggg — CHANGES_REQUESTED** [review]

**Review by yggg — APPROVED** [review]

**Review by yggg — APPROVED** [review]

**Pull request merged**